File: C#/SimplifyPath.cs
//Ian Hays
//02-06-2021
//https://leetcode.com/problems/simplify-path/
//SC: O(N^2) TC: O(N^2)

public class Solution {
    public string SimplifyPath(string path) {
        string[] directories = path.Trim('/').Split('/',StringSplitOptions.RemoveEmptyEntries).Where(x => x != ".").ToArray();
        Stack<string> stack = new Stack<string>();
        foreach(string directory in directories)
        {
            if(directory == "..")
            {
                if(stack.Count != 0) stack.Pop();
            } 
            else 
            {
                stack.Push(directory);
            }
        }
        return "/" +  String.Join("/", stack.Reverse());
    }
}

File: C++/SimplifyPath.cpp
//Ian Hays
//02-06-2021
//https://leetcode.com/problems/simplify-path/
//SC: O(N) TC: O(N)

class Solution {
public:
    string simplifyPath(string path) {
        stack<string> s;
        for(int i = 1; i < path.size(); i++){
            string tmp;
            while(i < path.size() && path[i] != '/'){
                tmp += path[i];
                i++;
            }
            if(tmp == ".."){
                if(!s.empty()) s.pop();
            } else {
                if(tmp != "." && tmp != "") s.push(tmp);
            }
        }
        
        string res;
        while(!s.empty()){
            string tmp = "/";
            tmp += s.top();
            s.pop();   
            res = tmp + res;
        }
        
        return res.empty() ? "/" : res;
    }
};





